components/ws-daemon/pkg/cpulimit/cfs_v2.go

```diff
@@ -0,0 +1,125 @@
+// Copyright (c) 2022 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package cpulimit
+
+import (
+	"bufio"
+	"math"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"golang.org/x/xerrors"
+)
+
+type CgroupV2CFSController string
+
+func (basePath CgroupV2CFSController) Usage() (CPUTime, error) {
+	usage, err := basePath.getFlatKeyedValue("usage_usec")
+	if err != nil {
+		return 0, err
+	}
+
+	return CPUTime(time.Duration(usage) * time.Microsecond), nil
+}
+
+func (basePath CgroupV2CFSController) SetLimit(limit Bandwidth) (changed bool, err error) {
+	quota, period, err := basePath.readCpuMax()
+	if err != nil {
+		return false, xerrors.Errorf("failed to read cpu max from %s: %w", basePath, err)
+	}
+
+	target := limit.Quota(period)
+	if quota == target {
+		return false, nil
+	}
+
+	err = basePath.writeQuota(target)
+	if err != nil {
+		return false, xerrors.Errorf("cannot set CFS quota of %d (period is %d, parent quota is %d): %w",
+			target.Microseconds(), period.Microseconds(), basePath.readParentQuota().Microseconds(), err)
+	}
+
+	return true, nil
+}
+
+func (basePath CgroupV2CFSController) NrThrottled() (uint64, error) {
+	throttled, err := basePath.getFlatKeyedValue("nr_throttled")
+	if err != nil {
+		return 0, err
+	}
+
+	return uint64(throttled), nil
+}
+
+func (basePath CgroupV2CFSController) readCpuMax() (time.Duration, time.Duration, error) {
+	cpuMaxPath := filepath.Join(string(basePath), "cpu.max")
+	cpuMax, err := os.ReadFile(cpuMaxPath)
+	if err != nil {
+		return 0, 0, xerrors.Errorf("unable to read cpu.max: %w", err)
+	}
+
+	parts := strings.Fields(string(cpuMax))
+	if len(parts) < 2 {
+		return 0, 0, xerrors.Errorf("cpu.max did not have expected number of fields: %s", parts)
+	}
+
+	var quota int64
```

Comment:
I think `quota` is always a positive value, right?
```suggestion
	var quota uint64
```